### DISCUSSION TITLE
How do we want to test *new* Rails applications?

### DISCUSSION THREAD

**Discussion by stevepolitodesign** [thoughtbot/suspenders#1294] [General]
In #1156 we [opted for RSpec over the default test framework](https://github.com/thoughtbot/suspenders/pull/1156#issuecomment-1985831463).

I think it's good to challenge this decision from time to time, since the decision wasn't unanimous.

Below are some questions that I think could guide this decision.

- What gains does RSpec provide in terms of **developer experience** that the default test framework does not provide?
- What gains does the default test framework provide in terms of **developer experience** that RSpec does not provide?
- What gains does RSpec provide in terms of **client success** that the default test framework does not provide?
- What gains does the default test framework provide in terms of **client success** that RSpec does not provide?

**Comment by DoodlingDev** [comment]
Just as another datapoint, I asked on Blue Sky during Rubyconf what test framework people use and where would they like to see thoughtbot's investment to Rails testing open source go, and it was unanimously "rspec". To the degree that it's "so default" no one ever thinks about it.

**Comment by JoelQ** [comment]
In terms of **developer experience**:
- 👍 RSpec provides better error messages
- 👍 RSpec provides better tools for backfilling tests on legacy code
- 🎨  spec-style writing seems more natural/less magical to me than repurposing methods in subclasses (I know that's controversial!)
- 🎨  greater flexibility of RSpec allows for many different styles of writing it, so less consistency from spec to spec
- 👎 greater flexibility of RSpec allows devs to play code golf with their specs
- 👎 default test framework provides a better parallel testing story out of the box (I hear)
- 👎 documentation for RSpec isn't great (not sure if default test framework is any better 🤷)

**Comment by stevepolitodesign** [comment]
> 👍 RSpec provides better tools for backfilling tests on legacy code

In the context of Suspenders, that may not be relevant, since we'd be generating a new application.

> 👎 documentation for RSpec isn't great (not sure if default test framework is any better 🤷)

There's a whole [section in the Guides](https://guides.rubyonrails.org/testing.html) dedicated to testing, which I personally think is valuable.

**Comment by nickcharlton** [comment]
I do really like RSpec (I'm with @JoelQ on all of his points).

> 👎 documentation for RSpec isn't great (not sure if default test framework is any better 🤷)

Yeah. I was thinking about this the other day, when I was, for the nth time going around in circles trying to find something I knew existed in some form but couldn't quite remember. I'm not sure how to go about it (as in: I don't know what improvements I'd make), but it's been bouncing around my head the last few weeks.

**Comment by stevepolitodesign** [comment]
I'm currently on a project using the default test suite provided by Rails, and I am finding I can accomplish a lot with [Minitest::Mock](https://www.rubydoc.info/gems/minitest/Minitest/Mock) and [#stub](https://www.rubydoc.info/gems/minitest/Object#stub-instance_method).
